### ISSUE TITLE
Authentication failed error

### ISSUE THREAD

**Issue by sqbell** [jar-o/rotvpn#1]
Hi! I am getting the following error:

```
(.env) [mbp rotvpn (master)]$ python3 rotvpn.py --name my-cool-vpn
SSH keys already seem to exist. Skipping generation.
Deploy rotvpn-my-cool-vpn not found.
Creating droplet ...
Droplet: rotvpn-my-cool-vpn, id=ID
in-progress
Waiting for server ... IP: None
IP Address: IP
Attempting to connect 0/5 ...
No valid connection. (Server probably not ready.)
Attempting to connect 1/5 ...
Traceback (most recent call last):
  File "rotvpn.py", line 28, in <module>
    provider.provision()
  File "/Users/user/Workspace/rotvpn/providers/digitalocean.py", line 94, in provision
    self.__install_wireguard()
  File "/Users/user/Workspace/rotvpn/providers/digitalocean.py", line 106, in __install_wireguard
    client.connect(self.ip_address, username='root')
  File "/Users/user/Workspace/rotvpn/.env/lib/python3.7/site-packages/paramiko/client.py", line 446, in connect
    passphrase,
  File "/Users/user/Workspace/rotvpn/.env/lib/python3.7/site-packages/paramiko/client.py", line 764, in _auth
    raise saved_exception
  File "/Users/user/Workspace/rotvpn/.env/lib/python3.7/site-packages/paramiko/client.py", line 740, in _auth
    self._transport.auth_publickey(username, key)
  File "/Users/user/Workspace/rotvpn/.env/lib/python3.7/site-packages/paramiko/transport.py", line 1580, in auth_publickey
    return self.auth_handler.wait_for_response(my_event)
  File "/Users/user/Workspace/rotvpn/.env/lib/python3.7/site-packages/paramiko/auth_handler.py", line 250, in wait_for_response
    raise e
paramiko.ssh_exception.AuthenticationException: Authentication failed.
```

**Comment by jar-o** [comment]
hm this line is suspicious

```
Droplet: rotvpn-my-cool-vpn, id=ID
```

`id=ID` seems like an issue maybe from the DigitalOcean API. when I ran just now I get

```
Droplet: rotvpn-my-cool-vpn, id=170161014
```

a DO droplet ID should always be a number. can you check the Droplet in the DO dashboard  and see if something is up? also perhaps check the key in the dashboard.

**Comment by sqbell** [comment]
@jar-o It had a real droplet ID, I just redacted it. I also redacted the IP from `IP Address`. Sorry - I should've stated it more clearly.

**Comment by jar-o** [comment]
ah I see. I wonder what the permissions on the `.key` files are? i.e. could it be a "permissions are too open" error, just not reported well? maybe the `chmod` in `digitalocean.py`  didn't work for some reason. perms should be `0600`

**Comment by sqbell** [comment]
It's 600 on the private key but 644 on the public:
```
[mbp rotvpn (master)]$ ls -la | grep key
-rw-------   1 user  staff  1678 Dec  4 11:33 rotvpn-digitalocean-private.key
-rw-r--r--   1 user  staff   380 Dec  4 11:33 rotvpn-digitalocean-public.key
```

Some debug logs:
```
DEBUG:paramiko.transport:starting thread (client mode): 0x55e4450
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.6.0
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.6p1)
DEBUG:paramiko.transport:kex algos:['curve25519-sha256', 'curve25519-sha256@libssh.org', 'ecdh-sha2-nistp256', 'ecdh-sha2-nistp384', 'ecdh-sha2-nistp521', 'diffie-hellman-group-exchange-sha256', 'diffie-hellman-group16-sha512', 'diffie-hellman-group18-sha512', 'diffie-hellman-group14-sha256', 'diffie-hellman-group14-sha1'] server key:['ssh-rsa', 'rsa-sha2-512', 'rsa-sha2-256', 'ecdsa-sha2-nistp256', 'ssh-ed25519'] client encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] server encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] client mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] server mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] client compress:['none', 'zlib@openssh.com'] server compress:['none', 'zlib@openssh.com'] client lang:[''] server lang:[''] kex follows?False
DEBUG:paramiko.transport:Kex agreed: curve25519-sha256@libssh.org
DEBUG:paramiko.transport:HostKey agreed: ssh-ed25519
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha2-256
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Adding ssh-ed25519 host key for 167.172.126.19: b'0a8e8343774158177b040fdff2a22650'
DEBUG:paramiko.transport:Trying discovered key b'aa80da821d29a89c337c669c7efc20d3' in /Users/user/.ssh/id_rsa
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (publickey) failed.
```

Also when I did the following it worked:
```diff
diff --git providers/digitalocean.py providers/digitalocean.py
index 6adcaa2..14e63f8 100644
--- providers/digitalocean.py
+++ providers/digitalocean.py
@@ -2,6 +2,7 @@
 import os, time, base64, json

 from Crypto.PublicKey import RSA
+import logging
 import paramiko
 from scp import SCPClient, SCPException
 import digitalocean
@@ -96,6 +97,8 @@ class DigitalOceanProvider:
         # NOTE there is no API specific way of telling if the SSH daemon is
         # ready. We just have to try in a loop
         time.sleep(10)
+        logging.basicConfig()
+        logging.getLogger("paramiko").setLevel(logging.DEBUG)
         key = paramiko.RSAKey.from_private_key_file(self.privkey_fn)
         client = paramiko.SSHClient()
         client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
@@ -103,7 +106,9 @@ class DigitalOceanProvider:
         for i in range(5):
             print("Attempting to connect {}/{} ...".format(i, 5))
             try:
-                client.connect(self.ip_address, username='root')
+                # import pdb; pdb.set_trace()
+                client.connect(self.ip_address, username='root', key_filename='/Users/user/Workspace/rotvpn/rotvpn-digitalocean-private.key')
+                # client.connect(self.ip_address, username='root', key_filename=key)
             except paramiko.ssh_exception.NoValidConnectionsError:
                 print('No valid connection. (Server probably not ready.)')
                 time.sleep(5)
```

**Comment by jar-o** [comment]
interesting, thanks for digging into to this. it seems like it found your default ssh key and used that?

```
DEBUG:paramiko.transport:Trying discovered key  in /Users/user/.ssh/id_rsa
```

oddly, on my system it does not do this, so I have struggled to reproduce. at any rate, there's no reason to not be more explicit. I've added `key_filename=` to the call.

https://github.com/jar-o/rotvpn/commit/762d793a65c9a01fb6936032e27769986b3e0fc6
